File: app/src/main/java/com/android/cast/dlna/demo/ControlFragment.java
package com.android.cast.dlna.demo;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.android.cast.dlna.DLNACastManager;
import com.android.cast.dlna.Utils;
import com.android.cast.dlna.control.ICastInterface;

import org.fourthline.cling.model.meta.Device;

public class ControlFragment extends Fragment implements IDisplayDevice, CastFragment.Callback {

    private TextView mPositionInfo;
    private SeekBar mPositionSeekBar;
    private TextView mVolumeInfo;
    private SeekBar mVolumeSeekBar;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_control, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initComponent(view);

        DLNACastManager.getInstance().registerActionCallbacks(
                new ICastInterface.CastEventListener() {
                    @Override
                    public void onSuccess(String result) {
                        Toast.makeText(getActivity(), "Cast: " + result, Toast.LENGTH_SHORT).show();
                        mCircleMsgHandler.start(0);
                    }

                    @Override
                    public void onFailed(String errMsg) {
                        Toast.makeText(getActivity(), errMsg, Toast.LENGTH_SHORT).show();
                    }
                },
                new ICastInterface.PlayEventListener() {
                    @Override
                    public void onSuccess(Void result) {
                        Toast.makeText(getActivity(), "Play", Toast.LENGTH_SHORT).show();
                    }

                    @Override
                    public void onFailed(String errMsg) {
                        Toast.makeText(getActivity(), errMsg, Toast.LENGTH_SHORT).show();
                    }
                },
                new ICastInterface.PauseEventListener() {
                    @Override
                    public void onSuccess(Void result) {
                        Toast.makeText(getActivity(), "Pause", Toast.LENGTH_SHORT).show();
                    }

                    @Override
                    public void onFailed(String errMsg) {
                        Toast.makeText(getActivity(), errMsg, Toast.LENGTH_SHORT).show();
                    }
                },
                new ICastInterface.StopEventListener() {
                    @Override
                    public void onSuccess(Void result) {
                        Toast.makeText(getActivity(), "Stop", Toast.LENGTH_SHORT).show();
                        mCircleMsgHandler.stop();
                    }

                    @Override
                    public void onFailed(String errMsg) {
                        Toast.makeText(getActivity(), errMsg, Toast.LENGTH_SHORT).show();
                    }
                },
                new ICastInterface.SeekToEventListener() {
                    @Override
                    public void onSuccess(Long result) {
                        Toast.makeText(getActivity(), "SeekTo: " + Utils.getStringTime(result), Toast.LENGTH_SHORT).show();
                    }

                    @Override
                    public void onFailed(String errMsg) {
                        Toast.makeText(getActivity(), errMsg, Toast.LENGTH_SHORT).show();
                    }
                }
        );
    }

    private void initComponent(View view) {
        mPositionInfo = view.findViewById(R.id.ctrl_position_info);

        view.findViewById(R.id.btn_cast).setOnClickListener(v -> new CastFragment().setCallback(this).show(getChildFragmentManager(), "CastFragment"));
        view.findViewById(R.id.btn_cast_pause).setOnClickListener(v -> DLNACastManager.getInstance().pause());
        view.findViewById(R.id.btn_cast_resume).setOnClickListener(v -> DLNACastManager.getInstance().play());
        view.findViewById(R.id.btn_cast_stop).setOnClickListener(v -> DLNACastManager.getInstance().stop());
        view.findViewById(R.id.btn_cast_mute).setOnClickListener(v -> DLNACastManager.getInstance().setMute(true));

        mPositionInfo = view.findViewById(R.id.ctrl_position_info);
        mPositionSeekBar = view.findViewById(R.id.ctrl_seek_position);
        mPositionSeekBar.setOnSeekBarChangeListener(mOnSeekBarChangeListener);
        mVolumeInfo = view.findViewById(R.id.ctrl_volume_info);
        mVolumeSeekBar = view.findViewById(R.id.ctrl_seek_volume);
        mVolumeSeekBar.setOnSeekBarChangeListener(mOnSeekBarChangeListener);
    }

    @Override
    public void onDestroyView() {
        DLNACastManager.getInstance().unregisterActionCallbacks();
        super.onDestroyView();
    }

    private final SeekBar.OnSeekBarChangeListener mOnSeekBarChangeListener = new SeekBar.OnSeekBarChangeListener() {
        @Override
        public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
        }

        @Override
        public void onStartTrackingTouch(SeekBar seekBar) {
        }

        @SuppressLint("NonConstantResourceId")
        @Override
        public void onStopTrackingTouch(SeekBar seekBar) {
            switch (seekBar.getId()) {
                case R.id.ctrl_seek_volume: {
                    DLNACastManager.getInstance().setVolume((int) (seekBar.getProgress() * 100f / seekBar.getMax()));
                    break;
                }
                case R.id.ctrl_seek_position: {
                    if (mDurationMillSeconds > 0) {
                        int position = (int) ((seekBar.getProgress() * 1f / seekBar.getMax()) * mDurationMillSeconds);
                        DLNACastManager.getInstance().seekTo(position);
                    }
                    break;
                }
            }
        }
    };

    private Device<?, ?, ?> mDevice;

    @Override
    public void setCastDevice(Device<?, ?, ?> device) {
        mDevice = device;
    }

    @Override
    public void onCastUrl(String url) {
        if (mDevice != null) {
            DLNACastManager.getInstance().cast(mDevice, CastObject.newInstance(url, Constants.CAST_ID, Constants.CAST_NAME));
        }
    }

    private long mDurationMillSeconds = 0;

    private final Runnable mRefreshUIRunnable = () -> {
        if (mDevice == null) return;
        DLNACastManager.getInstance().getPositionInfo(mDevice, (positionInfo, errMsg) -> {
            if (positionInfo != null) {
                mPositionInfo.setText(String.format("%s:%s", positionInfo.getRelTime(), positionInfo.getTrackDuration()));
                if (positionInfo.getTrackDurationSeconds() != 0) {
                    mDurationMillSeconds = positionInfo.getTrackDurationSeconds() * 1000;
                    mPositionSeekBar.setProgress((int) (positionInfo.getTrackElapsedSeconds() * 100 / positionInfo.getTrackDurationSeconds()));
                } else {
                    mPositionSeekBar.setProgress(0);
                }
            } else {
                mPositionInfo.setText(errMsg);
            }
        });

    };

    private final CircleMessageHandler mCircleMsgHandler = new CircleMessageHandler(1000, mRefreshUIRunnable);

    private static class CircleMessageHandler extends Handler {
        private static final int MSG = 101;

        private final long duration;
        private final Runnable runnable;

        public CircleMessageHandler(long duration, @NonNull Runnable runnable) {
            super(Looper.getMainLooper());
            this.duration = duration;
            this.runnable = runnable;
        }

        @Override
        public void handleMessage(@NonNull Message msg) {
            runnable.run();
            sendEmptyMessageDelayed(MSG, duration);
        }

        public void start(long delay) {
            stop();
            sendEmptyMessageDelayed(MSG, delay);
        }

        public void stop() {
            removeMessages(MSG);
        }
    }
}

File: app/src/main/java/com/android/cast/dlna/demo/ControlFragment.kt
package com.android.cast.dlna.demo

import android.content.Context
import android.media.AudioManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import com.android.cast.dlna.core.Utils
import com.android.cast.dlna.demo.CastFragment.Callback
import com.android.cast.dlna.demo.R.layout
import com.android.cast.dlna.dmc.DLNACastManager
import com.android.cast.dlna.dmc.control.ICastInterface
import com.android.cast.dlna.dmc.control.ICastInterface.CastEventListener
import com.android.cast.dlna.dmc.control.ICastInterface.GetInfoListener
import com.android.cast.dlna.dmc.control.ICastInterface.PauseEventListener
import com.android.cast.dlna.dmc.control.ICastInterface.PlayEventListener
import com.android.cast.dlna.dmc.control.ICastInterface.SeekToEventListener
import com.android.cast.dlna.dmc.control.ICastInterface.StopEventListener
import org.fourthline.cling.model.meta.Device
import org.fourthline.cling.support.model.PositionInfo
import org.fourthline.cling.support.model.TransportState
import java.util.UUID

class ControlFragment : Fragment(), IDisplayDevice, Callback {

    private val positionInfo: TextView? by lazy { view?.findViewById(R.id.ctrl_position_info) }
    private val positionSeekBar: SeekBar? by lazy { view?.findViewById(R.id.ctrl_seek_position) }
    private val volumeInfo: TextView? by lazy { view?.findViewById(R.id.ctrl_volume_info) }
    private val volumeSeekBar: SeekBar? by lazy { view?.findViewById(R.id.ctrl_seek_volume) }
    private val statusInfo: TextView? by lazy { view?.findViewById(R.id.ctrl_status_info) }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(layout.fragment_control, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initComponent(view)

        DLNACastManager.registerActionCallbacks(
            object : CastEventListener {
                override fun onSuccess(result: String) {
                    Toast.makeText(activity, "Cast: $result", Toast.LENGTH_LONG).show()
                    positionHandler.start()
                    mVolumeMsgHandler.start()
                }

                override fun onFailed(errMsg: String) {
                    Toast.makeText(activity, errMsg, Toast.LENGTH_LONG).show()
                }
            },
            object : PlayEventListener {
                override fun onSuccess(result: Void) {
                    Toast.makeText(activity, "Play", Toast.LENGTH_LONG).show()
                }

                override fun onFailed(errMsg: String) {
                    Toast.makeText(activity, errMsg, Toast.LENGTH_LONG).show()
                }
            },
            object : PauseEventListener {
                override fun onSuccess(result: Void) {
                    Toast.makeText(activity, "Pause", Toast.LENGTH_LONG).show()
                }

                override fun onFailed(errMsg: String) {
                    Toast.makeText(activity, errMsg, Toast.LENGTH_LONG).show()
                }
            },
            object : StopEventListener {
                override fun onSuccess(result: Void) {
                    Toast.makeText(activity, "Stop", Toast.LENGTH_LONG).show()
                    positionHandler.stop()
                    mVolumeMsgHandler.stop()
                }

                override fun onFailed(errMsg: String) {
                    Toast.makeText(activity, errMsg, Toast.LENGTH_LONG).show()
                }
            },
            object : SeekToEventListener {
                override fun onSuccess(result: Long) {
                    Toast.makeText(activity, "SeekTo: " + Utils.getStringTime(result), Toast.LENGTH_LONG).show()
                }

                override fun onFailed(errMsg: String) {
                    Toast.makeText(activity, errMsg, Toast.LENGTH_LONG).show()
                }
            }
        )
        DLNACastManager.registerSubscriptionListener(object : ICastInterface.ISubscriptionListener {
            override fun onSubscriptionTransportStateChanged(event: TransportState) {
                statusInfo?.text = event.value
            }
        })
    }

    private fun initComponent(view: View) {
        view.findViewById<View>(R.id.btn_cast).setOnClickListener { CastFragment(this).show(childFragmentManager) }
        view.findViewById<View>(R.id.btn_cast_pause).setOnClickListener { DLNACastManager.pause() }
        view.findViewById<View>(R.id.btn_cast_resume).setOnClickListener { DLNACastManager.play() }
        view.findViewById<View>(R.id.btn_cast_stop).setOnClickListener { DLNACastManager.stop() }
        view.findViewById<View>(R.id.btn_cast_mute).setOnClickListener { DLNACastManager.setMute(true) }
        positionSeekBar?.setOnSeekBarChangeListener(seekBarChangeListener)
        volumeSeekBar?.setOnSeekBarChangeListener(seekBarChangeListener)
    }

    override fun onDestroyView() {
        DLNACastManager.unregisterActionCallbacks()
        super.onDestroyView()
    }

    private val seekBarChangeListener: OnSeekBarChangeListener = object : OnSeekBarChangeListener {
        override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {}
        override fun onStartTrackingTouch(seekBar: SeekBar) {}

        override fun onStopTrackingTouch(seekBar: SeekBar) {
            when (seekBar.id) {
                R.id.ctrl_seek_volume -> {
                    DLNACastManager.setVolume((seekBar.progress * 100f / seekBar.max).toInt())
                }

                R.id.ctrl_seek_position -> {
                    if (durationMillSeconds > 0) {
                        val position = (seekBar.progress * 1f / seekBar.max * durationMillSeconds).toInt()
                        DLNACastManager.seekTo(position.toLong())
                    }
                }
            }
        }
    }

    private var device: Device<*, *, *>? = null

    override fun setCastDevice(device: Device<*, *, *>?) {
        this.device = device
        if (device == null) {
            positionInfo?.text = ""
            positionSeekBar?.progress = 0
            volumeInfo?.text = ""
            volumeSeekBar?.progress = 0
            positionHandler.stop()
            mVolumeMsgHandler.stop()
        }
        // reconnect device, should recover status?
    }

    override fun onCastUrl(url: String?) {
        if (device != null) {
            DLNACastManager.cast(device!!, CastObject.newInstance(url!!, UUID.randomUUID().toString(), "Test Sample"))
        }
    }

    private var durationMillSeconds: Long = 0

    private val positionHandler = CircleMessageHandler(1000, Runnable {
        if (device == null) return@Runnable
        // update position text and progress
        DLNACastManager.getPositionInfo(device!!, object : GetInfoListener<PositionInfo> {
            override fun onGetInfoResult(t: PositionInfo?, errMsg: String?) {
                if (t != null) {
                    this@ControlFragment.positionInfo?.text = String.format("%s/%s", t.relTime, t.trackDuration)
                    if (t.trackDurationSeconds != 0L) {
                        durationMillSeconds = t.trackDurationSeconds * 1000
                        positionSeekBar?.progress = (t.trackElapsedSeconds * 100 / t.trackDurationSeconds).toInt()
                    } else {
                        positionSeekBar?.progress = 0
                    }
                } else {
                    this@ControlFragment.positionInfo?.text = errMsg
                }
            }
        })
    })

    private val mVolumeMsgHandler = CircleMessageHandler(3000, Runnable {
        if (device == null) return@Runnable
        // update volume
        DLNACastManager.getVolumeInfo(device!!, object : GetInfoListener<Int> {
            override fun onGetInfoResult(t: Int?, errMsg: String?) {
                if (t != null && activity != null) {
                    val audioManager = requireActivity().getSystemService(Context.AUDIO_SERVICE) as AudioManager
                    val maxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
                    volumeSeekBar?.progress = t
                    volumeInfo?.text = String.format("%s/%s", (t / 100f * maxVolume).toInt(), maxVolume)
                } else {
                    volumeInfo?.text = errMsg
                }
            }

        })
    })

    // ------------------------------------------------------------
    // ---- Handler
    // ------------------------------------------------------------
    private class CircleMessageHandler(
        private val duration: Long,
        private val runnable: Runnable
    ) : Handler(Looper.getMainLooper()) {
        override fun handleMessage(msg: Message) {
            runnable.run()
            sendEmptyMessageDelayed(MSG, duration)
        }

        fun start(delay: Long = 0L) {
            stop()
            sendEmptyMessageDelayed(MSG, delay)
        }

        fun stop() {
            removeMessages(MSG)
        }

        companion object {
            private const val MSG = 101
        }
    }
}